---
title: "vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


# ClusterGVis

To enhance clustering and visualization of time-series gene expression data from RNA-Seq experiments, we present the ClusterGVis package. This tool enables concise and elegant analysis of time-series gene expression data in a simple, one-step operation. Additionally, you can perform enrichment analysis for each cluster using the enrichCluster function, which integrates seamlessly with clusterProfiler. ClusterGVis empowers you to create publication-quality visualizations with ease.

Comprehensive documentation can be found <https://junjunlab.github.io/ClusterGvis-manual>.

## Installation

```{r,eval=TRUE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("ClusterGVis")
```

## Usage

### Basic examples:

```{r,eval=TRUE}
library(SummarizedExperiment)
library(S4Vectors)
library(ClusterGVis)

# a data.frame or SummarizedExperiment object
data("exps")

head(exps)

# using kemans for clustering
ck <- clusterData(obj = exps,
                  cluster.method = "kmeans",
                  cluster.num = 8)

# construct a SummarizedExperiment object
sce <- SummarizedExperiment(assays = list(counts = exps),
                            colData = S4Vectors::DataFrame(sample = colnames(exps),
                                                           row.names = colnames(exps))
                            )

sce

# check suitable cluster nmbers
getClusters(obj = sce)

# using kemans for clustering
ck2 <- clusterData(obj = sce,
                  cluster.method = "kmeans",
                  cluster.num = 8)

# plot line only
visCluster(object = ck,
           plot.type = "line")

# plot heatmap only
visCluster(object = ck,
           plot.type = "heatmap")
```

### Integration with seurat object:

```{r,eval=TRUE}
library(Seurat)

data("pbmc_subset")

# find markers for every cluster compared to all remaining cells
# report only the positive ones
pbmc.markers.all <- Seurat::FindAllMarkers(pbmc_subset,
                                           only.pos = TRUE,
                                           min.pct = 0.25,
                                           logfc.threshold = 0.25)

# get top 10 genes
pbmc.markers <- pbmc.markers.all %>%
  dplyr::group_by(cluster) %>%
  dplyr::top_n(n = 20, wt = avg_log2FC)

# check
head(pbmc.markers)

# prepare data from seurat object
st.data <- prepareDataFromscRNA(object = pbmc_subset,
                                diffData = pbmc.markers,
                                showAverage = TRUE)

# check
str(st.data)

# line plot
visCluster(object = st.data,
           plot.type = "line")

# add gene name
markGenes = unique(pbmc.markers$gene)[sample(1:length(unique(pbmc.markers$gene)),40,
                                             replace = F)]

# heatmap plot
pdf('sc1.pdf',height = 10,width = 6,onefile = F)
visCluster(object = st.data,
           plot.type = "heatmap",
           column_names_rot = 45,
           markGenes = markGenes,
           cluster.order = c(1:9))
dev.off()
```

# Session Info

```{r}
sessionInfo()
```
